In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import (
    RandomizedSearchCV,
    cross_val_score,
    cross_validate,
    train_test_split,
)
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler
from sklearn.compose import (
    ColumnTransformer,
    TransformedTargetRegressor,
    make_column_transformer,
)

# Load data

In [2]:
data = pd.read_csv('../data/cleaned_data_v2.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 47 columns):
 #   Column                                                                                                                                                                                                                         Non-Null Count  Dtype  
---  ------                                                                                                                                                                                                                         --------------  -----  
 0   case number                                                                                                                                                                                                                    696 non-null    object 
 1   Who was the member adjudicating the decision?                                                                                         

In [3]:
data = data.rename(columns={
    'What was the length of the tenancy, or in other words, how long had the tenants lived at the residence in question? '
    :'What was the length of the tenancy',
    'What was the amount of the rental deposit? ':'rental deposit amount',
    'If any rent increases occurred, what was the rent after the increase(s)?':'was there an rent increases',
    'Over how many months did the arrears accumulate? ':'Over how many months did the arrears accumulate?',
    'If the tenant made a payment on the arrears after the eviction notice was served and/or prior to the hearing, what was the amount of the payment? ':'Does the tenant made a payment on the arrears after the eviction notice',
    'What were the specific mental, medical, or physical conditions of the tenant, if any? ':'What were the specific mental, medical, or physical conditions of the tenant, if any?'})

In [4]:
train_df, test_df = train_test_split(data, test_size=0.2, random_state=123)
X_train, y_train = train_df.drop(columns = ["What was the outcome of the case?"]), train_df["What was the outcome of the case?"]
X_test, y_test = test_df.drop(columns = ["What was the outcome of the case?"]), test_df["What was the outcome of the case?"]
X_train.head()

case number Who was the member adjudicating the decision?  \
195  TSL-90833-17                               Roger Rodrigues   
553  TNL-07861-18                                  Nancy Morris   
598  TSL-96267-18                                David Mungovan   
645  TEL-81094-17                               Shelby Whittick   
634  TEL-80073-17                                  Jim McMaster   

    What was the location of the landlord tenant board?  \
195                                            Toronto    
553                                            Toronto    
598                                            Toronto    
645                                             Whitby    
634                                            Toronto    

     Did the decision state the landlord was represented?  \
195                                                  1      
553                                                  0      
598                                                  1      
645                                                  0      
634                                                  1      

     Did the decision state the landlord attended the hearing?  \
195                                                  0           
553                                                  1           
598                                                  0           
645                                                  1           
634                                                  1           

     Did the decision state the tenant was represented?  \
195                                                  0    
553                                                  0    
598                                                  0    
645                                                  0    
634                                                  0    

     Did the decision state the tenant attended the hearing?  \
195                                                  0         
553                                                  1         
598                                                  1         
645                                                  1         
634                                                  1         

     Did the decision state the landlord was a not-for-profit landlord (e.g. Toronto Community Housing)?  \
195                                                  0                                                     
553                                                  0                                                     
598                                                  0                                                     
645                                                  0                                                     
634                                                  0                                                     

     Did the decision state the tenant was collecting a subsidy?  \
195                                                  0             
553                                                  0             
598                                                  0             
645                                                  0             
634                                                  0             

     What was the length of the tenancy  ...  L9 present?  N5 present?  \
195                            6.306049  ...            0            0   
553                           24.000000  ...            0            0   
598                            6.306049  ...            0            0   
645                            6.306049  ...            0            0   
634                            6.306049  ...            0            0   

     N6 present?  N7 present?  N8 present?  T1 present?  T2 present?  \
195            0            0            0            0            0   
553            0            0            0            0            0   
598            0         

In [5]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 556 entries, 195 to 510
Data columns (total 46 columns):
 #   Column                                                                                                                                                                                                                         Non-Null Count  Dtype  
---  ------                                                                                                                                                                                                                         --------------  -----  
 0   case number                                                                                                                                                                                                                    556 non-null    object 
 1   Who was the member adjudicating the decision?                                                                                       

In [6]:
y_train.value_counts()

No relief                   352
Postponement of eviction    124
Payment plan                 64
Relief                       12
Conditional Order             4
Name: What was the outcome of the case?, dtype: int64

## Preprocess data with pipeline

In [7]:
categorical_features = [
    'Who was the member adjudicating the decision?',
    'What was the location of the landlord tenant board?'
]

numeric_features = [
    'What was the length of the tenancy',
    'What was the monthly rent?',
    'rental deposit amount',
    'What was the total amount of arrears?',
    'Over how many months did the arrears accumulate?'
]

ordinal_features = [
    'Was the tenant employed at the time of the hearing?',
    'If the tenant was not employed, did the decision state the tenant was receiving any form of government assistance (e.g. OW, childcare benefits, ODSP, OSAP)?',
    'If the tenant was employed, did the decision state any doubts about the stability of employment e.g. lack of guaranteed hours, contract work, etc.?',
    'Payment Plan'
]

drop_features = ['case number']

ordering_ordinal = [
    [-1,0,1],
    [-1,0,1],
    [-1,0,1],
    [0, 0.5, 1]
]

passthrough_feats = list(
    set(X_train.columns)
    - set(categorical_features)
    - set(numeric_features)
    - set(ordinal_features)
    - set(drop_features)
)

In [8]:
preprocessor = make_column_transformer(
        (OneHotEncoder(handle_unknown="ignore"),categorical_features),
        ("passthrough", passthrough_feats),
        (StandardScaler(),numeric_features),
        (OrdinalEncoder(categories = ordering_ordinal),ordinal_features),
        ("drop",drop_features))
preprocessor

ColumnTransformer(transformers=[('onehotencoder',
                                 OneHotEncoder(handle_unknown='ignore'),
                                 ['Who was the member adjudicating the '
                                  'decision?',
                                  'What was the location of the landlord '
                                  'tenant board?']),
                                ('passthrough', 'passthrough',
                                 ['Did the member find the tenant had or '
                                  'seemed to have the ability to pay rent, but '
                                  'chose not do so?',
                                  'Did the decisions state postponement would '
                                  'result in the t...
                                 ['Was the tenant employed at the time of the '
                                  'hearing?',
                                  'If the tenant was not employed, did the '
                                  'decision state the tenant was receiving any '
                                  'form of government assistance (e.g. OW, '
                                  'childcare benefits, ODSP, OSAP)?',
                                  'If the tenant was employed, did the '
                                  'decision state any doubts about the '
                                  'stability of employment e.g. lack of '
                                  'guaranteed hours, contract work, etc.?',
                                  'Payment Plan']),
                                ('drop', 'drop', ['case number'])])

## Baseline model

In [9]:
# adapt from DSCI573 lecture note
def mean_std_cross_val_scores(model, X_train, y_train, **kwargs):
    """
    Returns mean and std of cross validation

    Parameters
    ----------
    model :
        scikit-learn model
    X_train : numpy array or pandas DataFrame
        X in the training data
    y_train :
        y in the training data

    Returns
    ----------
        pandas Series with mean scores from cross_validation
    """

    scores = cross_validate(model, X_train, y_train, **kwargs)

    mean_scores = pd.DataFrame(scores).mean()
    std_scores = pd.DataFrame(scores).std()
    out_col = []

    for i in range(len(mean_scores)):
        out_col.append((f"%0.3f (+/- %0.3f)" % (mean_scores[i], std_scores[i])))

    return pd.Series(data=out_col, index=mean_scores.index)

In [10]:
from sklearn.dummy import DummyClassifier

results_dict = {}
scoring_metrics = ["accuracy"]

dummy = make_pipeline(preprocessor, DummyClassifier(random_state = 123))
results_dict["dummy"] = mean_std_cross_val_scores(
    dummy, X_train, y_train, scoring = scoring_metrics, return_train_score=True
)

results_df = pd.DataFrame(results_dict).T
results_df

/Users/AliceYuuu/opt/miniconda3/lib/python3.10/site-packages/sklearn/model_selection/_split.py:684: UserWarning: The least populated class in y has only 4 members, which is less than n_splits=5.
  warnings.warn(


fit_time         score_time      test_accuracy  \
dummy  0.011 (+/- 0.004)  0.007 (+/- 0.003)  0.633 (+/- 0.004)   

          train_accuracy  
dummy  0.633 (+/- 0.001)

_With no tuned hyperparameters of DummyClassifier, the mean cross validation score of our baseline model is 0.633. Since it is a a multi-class classification and there is class imbalance, this accuracy score is low and not quite representative._

# LightGBM model

In [11]:
from lightgbm.sklearn import LGBMClassifier

pipe_lgbm = make_pipeline(preprocessor, LGBMClassifier(random_state=123))

results_dict["lgbm"] = mean_std_cross_val_scores(
    pipe_lgbm, X_train, y_train, scoring = scoring_metrics, return_train_score=True
)

results_df = pd.DataFrame(results_dict).T
results_df

/Users/AliceYuuu/opt/miniconda3/lib/python3.10/site-packages/sklearn/model_selection/_split.py:684: UserWarning: The least populated class in y has only 4 members, which is less than n_splits=5.
  warnings.warn(


fit_time         score_time      test_accuracy  \
dummy  0.011 (+/- 0.004)  0.007 (+/- 0.003)  0.633 (+/- 0.004)   
lgbm   0.211 (+/- 0.030)  0.009 (+/- 0.002)  0.750 (+/- 0.029)   

          train_accuracy  
dummy  0.633 (+/- 0.001)  
lgbm   1.000 (+/- 0.000)

_With default hyperparameters, our LightGBM model gets a better mean cross validation score (0.75) than the score of base model (0.663). However, we can see the model is very overfit, with the mean train score being 0.25 higher than the mean cv score._

# Hyperparameter optimization

In [12]:
param_grid = {
    "lgbmclassifier__learning_rate": np.arange(0.01, 0.2, 0.001),
    "lgbmclassifier__max_depth": np.arange(0, 12, 1),
    "lgbmclassifier__num_leaves": np.arange(10, 80, 2),
    "lgbmclassifier__class_weight": ["balanced", None]
}

random_search = RandomizedSearchCV(
    pipe_lgbm, 
    param_grid, 
    n_jobs=-1, 
    n_iter=30, 
    random_state=123,
    scoring=scoring_metrics, 
    refit='accuracy',
    return_train_score=True
)

random_search.fit(X_train, y_train)

results = pd.DataFrame(random_search.cv_results_)[
    [
        "mean_test_accuracy",
        "param_lgbmclassifier__learning_rate",
        "param_lgbmclassifier__max_depth",
        "param_lgbmclassifier__num_leaves",
        "param_lgbmclassifier__class_weight",
        "mean_fit_time",
        "rank_test_accuracy",
    ]
].set_index("rank_test_accuracy").sort_index().T
results

/Users/AliceYuuu/opt/miniconda3/lib/python3.10/site-packages/sklearn/model_selection/_split.py:684: UserWarning: The least populated class in y has only 4 members, which is less than n_splits=5.
  warnings.warn(


rank_test_accuracy                         1         2         3         4   \
mean_test_accuracy                   0.778797  0.775209   0.76982  0.762629   
param_lgbmclassifier__learning_rate     0.117     0.109      0.13     0.017   
param_lgbmclassifier__max_depth             1         1         1         9   
param_lgbmclassifier__num_leaves           14        36        52        58   
param_lgbmclassifier__class_weight       None      None      None      None   
mean_fit_time                        0.054511  0.058014  0.059689  0.207768   

rank_test_accuracy                         4         6         7         8   \
mean_test_accuracy                   0.762629  0.760827  0.755421  0.755405   
param_lgbmclassifier__learning_rate     0.017     0.105     0.127     0.187   
param_lgbmclassifier__max_depth             7         5         5         2   
param_lgbmclassifier__num_leaves           64        72        40        66   
param_lgbmclassifier__class_weight       None      None      None      None   
mean_fit_time                        0.182239  0.151301  0.142716  0.084502   

rank_test_accuracy                         9         10  ...        21  \
mean_test_accuracy                   0.751802   0.74823  ...  0.732062   
param_lgbmclassifier__learning_rate     0.195      0.12  ...     0.082   
param_lgbmclassifier__max_depth             5         0  ...         0   
param_lgbmclassifier__num_leaves           56        16  ...        40   
param_lgbmclassifier__class_weight       None  balanced  ...  balanced   
mean_fit_time                        0.133934  0.198323  ...  0.444405   

rank_test_accuracy                         22        23        24        25  \
mean_test_accuracy                   0.726657  0.719434  0.714157  0.714077   
param_lgbmclassifier__learning_rate     0.118     0.166     0.047     0.066   
param_lgbmclassifier__max_depth             4         3         5         5   
param_lgbmclassifier__num_leaves           24        24        30        76   
param_lgbmclassifier__class_weight   balanced  balanced  balanced  balanced   
mean_fit_time                        0.121146  0.128563   0.15111  0.150672   

rank_test_accuracy                         26        27        28        29  \
mean_test_accuracy                   0.694369  0.683591  0.678201  0.661969   
param_lgbmclassifier__learning_rate     0.025     0.019     0.015     0.052   
param_lgbmclassifier__max_depth             9        11         9         2   
param_lgbmclassifier__num_leaves           76        70        36        50   
param_lgbmclassifier__class_weight   balanced  balanced  balanced  balanced   
mean_fit_time                        0.307662   0.26708  0.217404  0.085321   

rank_test_accuracy                         30  
mean_test_accuracy                   0.609733  
param_lgbmclassifier__learning_rate     0.029  
param_lgbmclassifier__max_depth             1  
param_lgbmclassifier__num_leaves           36  
param_lgbmclassifier__class_weight   balanced  
mean_fit_time                         0.05708  

[6 rows x 30 columns]

In [13]:
best_pipe_lgbm = make_pipeline(
    preprocessor, 
    LGBMClassifier(num_leaves = 14, learning_rate = 0.117, max_depth = 1, class_weight=None, random_state=123)
)

In [14]:
results_dict["best_lgbm"] = mean_std_cross_val_scores(
    best_pipe_lgbm, X_train, y_train, scoring = scoring_metrics, return_train_score=True
)

results_df = pd.DataFrame(results_dict).T
results_df

/Users/AliceYuuu/opt/miniconda3/lib/python3.10/site-packages/sklearn/model_selection/_split.py:684: UserWarning: The least populated class in y has only 4 members, which is less than n_splits=5.
  warnings.warn(


fit_time         score_time      test_accuracy  \
dummy      0.011 (+/- 0.004)  0.007 (+/- 0.003)  0.633 (+/- 0.004)   
lgbm       0.211 (+/- 0.030)  0.009 (+/- 0.002)  0.750 (+/- 0.029)   
best_lgbm  0.043 (+/- 0.008)  0.008 (+/- 0.001)  0.779 (+/- 0.044)   

              train_accuracy  
dummy      0.633 (+/- 0.001)  
lgbm       1.000 (+/- 0.000)  
best_lgbm  0.816 (+/- 0.006)

_As shown above, We tuned three hyperparameters: learning_rate, max_depth, num_leaves and class_weight. The highest mean cv score we got is 0.779 with learning_rate=0.117, max_depth=1, num_leaves=14 and class_weight=None. This is better than with the default hyperparameter which has a mean cv score of 0.75. Additionally, we can see that the model is not overfit anymore with the mean train score being only 0.037 higher than the mean cv score._

# Interpretation and feature importances 

In [15]:
#!pip install eli5

In [24]:
import eli5
# eli5 feature importance weights for LGBM classifier
categorical_feature_names = (
    random_search.best_estimator_
    .named_steps["columntransformer"]
    .named_transformers_["onehotencoder"]
    .get_feature_names_out(categorical_features)
    .tolist()
)

feature_names = (
    categorical_feature_names + passthrough_feats + numeric_features + ordinal_features
)

best_pipe_lgbm.fit(X_train, y_train)

eli5.explain_weights(
    best_pipe_lgbm.named_steps["lgbmclassifier"], top=None, feature_names=feature_names
)

Explanation(estimator='LGBMClassifier(learning_rate=0.117, max_depth=1, num_leaves=14,\n               random_state=123)', description='\nLightGBM feature importances; values are numbers 0 <= x <= 1;\nall values sum to 1.\n', error=None, method='feature importances', is_regression=False, targets=None, feature_importances=FeatureImportances(importances=[FeatureWeight(feature='Payment Plan', weight=0.3699344674776561, std=None, value=None), FeatureWeight(feature='Did the decision state the tenant attended the hearing?', weight=0.13823664109891493, std=None, value=None), FeatureWeight(feature='What was the total amount of arrears?', weight=0.08020876826330023, std=None, value=None), FeatureWeight(feature='Did the decision mention the tenant’s difficulty finding alternative housing for any reason e.g.physical limitations, reliance on social assistance, etc.?', weight=0.07925797979126965, std=None, value=None), FeatureWeight(feature='Over how many months did the arrears accumulate?', weight=0.036064750245281114, std=None, value=None), FeatureWeight(feature='Did the member find the tenant had sufficient income to pay rent?', weight=0.0339236149550832, std=None, value=None), FeatureWeight(feature='rental deposit amount', weight=0.027448368976710654, std=None, value=None), FeatureWeight(feature='Did the decisions state postponement would result in the tenant accruing additional arrears?', weight=0.026897076995988985, std=None, value=None), FeatureWeight(feature='What were the specific mental, medical, or physical conditions of the tenant, if any?', weight=0.02613466643901641, std=None, value=None), FeatureWeight(feature='Did the decision state the tenant was represented?', weight=0.025998538068112583, std=None, value=None), FeatureWeight(feature='Did the decision state the tenant was collecting a subsidy?', weight=0.02494393290969799, std=None, value=None), FeatureWeight(feature='N8 present?', weight=0.01751787642705849, std=None, value=None), FeatureWeight(feature='If the tenant was employed, did the decision state any doubts about the stability of employment e.g. lack of guaranteed hours, contract work, etc.?', weight=0.015527109214551225, std=None, value=None), FeatureWeight(feature='What was the location of the landlord tenant board?_London', weight=0.012399397644851312, std=None, value=None), FeatureWeight(feature='What was the length of the tenancy', weight=0.011993597329259685, std=None, value=None), FeatureWeight(feature='Who was the member adjudicating the decision?_Ruth Carey', weight=0.01168787464792816, std=None, value=None), FeatureWeight(feature='What was the monthly rent?', weight=0.011199042790842028, std=None, value=None), FeatureWeight(feature='Did the decision mention any other extenuating circumstances experienced by the tenant leading up to or during the period of the claim (e.g. hospitalization, death in the family, etc.)?', weight=0.00909739641317723, std=None, value=None), FeatureWeight(feature='Did the member find the tenant had or seemed to have the ability to pay rent, but chose not do so?', weight=0.007427757505106211, std=None, value=None), FeatureWeight(feature='What was the location of the landlord tenant board?_Whitby', weight=0.005620003883813954, std=None, value=None), FeatureWeight(feature='N5 present?', weight=0.0045223287619862865, std=None, value=None), FeatureWeight(feature='Did the decision state the landlord was represented?', weight=0.0035214496899013295, std=None, value=None), FeatureWeight(feature='Who was the member adjudicating the decision?_Harry Cho', weight=0.0034288614149685956, std=None, value=None), FeatureWeight(feature='Does the tenant made a payment on the arrears after the eviction notice', weight=0.003202273126179162, std=None, value=None), FeatureWeight(feature='L2 present?', weight=0.0025460350082574354, std=None, value=None), FeatureWeight(feature='What was the location of the landlord tenant board?_Toronto', weight=0.0025224894774488306, std=None, value=None), FeatureWeight(

_Based on the eli5 results above, it seems that "Payment Plan" is the strongest variable used by the LGBM Classifiers with a value of 0.3699. The second strongest variable "Did the decision state the tenant attended the hearing?" gets a weight of 0.1382. All other variables have a weight lower than 0.081, which indicates they are not as important._

# Conclusion

In [25]:
test_score = best_pipe_lgbm.score(X_test, y_test)
test_score

0.7857142857142857

_The test score for my LightGBM model is 0.786, which agrees with the mean validation score of 0.779 from before. Since the validation score and the test score are similar, I don't think there is optimization bias. If there is optimization bias, the validation score would be higher than the test score. Therefore, I have decent trust in my model, and believe it would do well to deployment data._